src/tests/containerizer/cgroups2_tests.cpp

```diff
@@ -210,6 +220,148 @@ TEST_F(Cgroups2Test, ROOT_CGROUPS2_EnableAndDisable)
+        SAFE_EXIT(EXIT_FAILURE, "Expected allowed read to succeed");
+      }
+    }
+    foreach(const OpenArgs& args, blockedAccesses) {
+      if (os::open(args.first, args.second).isSome()) {
+        SAFE_EXIT(EXIT_FAILURE, "Expected blocked read to fail");
+      }
+    }
+
+    ASSERT_SOME(ebpf::cgroups2::detach(path, attached->at(0)));
+
+    // Check that we can do both the "allowedAccesses" and "blockedAccesses".
+    foreach(const OpenArgs& args, allowedAccesses) {
+      if (os::open(args.first, args.second).isError()) {
+        SAFE_EXIT(EXIT_FAILURE, "Expected successful read after detaching"
+          " device controller program");
+      }
+    }
+    foreach(const OpenArgs& args, blockedAccesses) {
+      if (os::open(args.first, args.second).isError()) {
+        SAFE_EXIT(EXIT_FAILURE, "Expected successful read after detaching"
+          " device controller program");
+      }
+    }
+
+    ::_exit(EXIT_SUCCESS);
+  }
+
+  AWAIT_EXPECT_WEXITSTATUS_EQ(EXIT_SUCCESS, process::reap(pid));
+}
+
+
+INSTANTIATE_TEST_CASE_P(
+    DeviceControllerTestParams,
+    DeviceControllerTestFixture,
+    ::testing::Values<DeviceControllerTestParams>(
+      DeviceControllerTestParams{
+            vector<devices::Entry>{},
+            vector<devices::Entry>{},
+            vector<OpenArgs>{},
+            // Block accesses by default.
+            vector<OpenArgs>{{os::DEV_NULL, O_RDWR}, {os::DEV_NULL, O_RDWR}}
+       },
+       DeviceControllerTestParams{
+            vector<devices::Entry>{},
+            vector<devices::Entry>{*devices::Entry::parse("c 1:3 rwm")},
+            vector<OpenArgs>{},
+            vector<OpenArgs>{{os::DEV_NULL, O_RDWR}}
+        },
+        DeviceControllerTestParams{
+            // allow /dev/null
+            vector<devices::Entry>{*devices::Entry::parse("c 1:3 rwm")},
+            vector<devices::Entry>{},
+            // read-write allowed
+            vector<OpenArgs>{{os::DEV_NULL, O_RDWR}},
+            vector<OpenArgs>{}
+        },
+        DeviceControllerTestParams{
+            // allow /dev/null
+            vector<devices::Entry>{*devices::Entry::parse("c 1:3 r")},
+            vector<devices::Entry>{},
+            // read-only allowed
+            vector<OpenArgs>{{os::DEV_NULL, O_RDONLY}},
+            // read-write is blocked
+            vector<OpenArgs>{{os::DEV_NULL, O_RDWR}}
+        },
+        DeviceControllerTestParams{
+            vector<devices::Entry>{*devices::Entry::parse("c 1:3 w")},
+            vector<devices::Entry>{},
+            // write-only allowed
+            vector<OpenArgs>{{os::DEV_NULL, O_WRONLY}},
+            // read is blocked
+            vector<OpenArgs>{{os::DEV_NULL, O_RDWR}, {os::DEV_NULL, O_RDONLY}}
+        },
+        DeviceControllerTestParams{
+            vector<devices::Entry>{*devices::Entry::parse("b 1:3 rwm")},
+            vector<devices::Entry>{},
+            vector<OpenArgs>{},
+            // /dev/null is blocked
```

Comment:
only because the entry has the wrong device type though, which isn't very clear here